tests/test_pmem_allocator.cpp

```diff
@@ -51,56 +52,62 @@ TEST_F(EnginePMemAllocatorTest, TestBasicAlloc) {
-      remove(pmem_path.c_str());
-      PMEMAllocator* pmem_alloc = PMEMAllocator::NewPMEMAllocator(
-          pmem_path, pmem_size, num_segment_block, block_size, num_thread, true,
-          false, nullptr);
-      ASSERT_NE(pmem_alloc, nullptr);
-      // Test function.
-      auto TestPmemAlloc = [&](size_t) {
-        thread_manager_->MaybeInitThread(access_thread);
-        std::vector<SpaceEntry> records;
-        uint64_t kvpairs = pmem_size / block_size;
-        for (uint64_t j = 0; j < kvpairs; ++j) {
-          auto space_entry = pmem_alloc->Allocate(alloc_size);
-          records.push_back(space_entry);
+      for (auto num_thread : num_threads) {
+        // init pmem allocator and thread_manager.
+        thread_manager_.reset(new ThreadManager(num_thread));
+        PMEMAllocator* pmem_alloc = PMEMAllocator::NewPMEMAllocator(
+            pmem_path, pmem_size, num_segment_block, block_size, num_thread,
+            true, false, nullptr);
+        if (block_size * num_segment_block * num_thread > pmem_size) {
+          ASSERT_EQ(pmem_alloc, nullptr);
+          continue;
+        } else {
+          ASSERT_NE(pmem_alloc, nullptr);
         }
 
-        auto alloc_bytes = pmem_alloc->PMemUsageInBytes();
-
-        for (uint64_t j = 0; j < records.size(); ++j) {
-          pmem_alloc->Free(records[j]);
-        }
+        // Test function: allocate all pmem, and free all under multi-threaded
+        // scenario.
+        auto TestPmemAlloc = [&](size_t) {
+          thread_manager_->MaybeInitThread(access_thread);
+          std::vector<SpaceEntry> records;
+          for (uint64_t j = 0; j < num_segment_block; ++j) {
+            auto space_entry = pmem_alloc->Allocate(alloc_size);
+            records.push_back(space_entry);
+          }
+          for (uint64_t j = 0; j < records.size(); ++j) {
+            pmem_alloc->Free(records[j]);
+          }
+        };
+        LaunchNThreads(num_thread, TestPmemAlloc);
 
         ASSERT_EQ(pmem_alloc->PMemUsageInBytes(), 0LL);
-        records.clear();
+        pmem_alloc->BackgroundWork();
+        access_thread.Release();
 
-        uint64_t alloc_cnt = 1;
-        // again allocate pmem
+        // Then allocate all pmem.
+        thread_manager_->MaybeInitThread(access_thread);
+        int alloc_cnt = 0;
         while (true) {
-          auto space_entry = pmem_alloc->Allocate(alloc_size);
-          ASSERT_NE(space_entry.size, 0);
-          if (pmem_alloc->PMemUsageInBytes() == alloc_bytes) break;
+          SpaceEntry space_entry = pmem_alloc->Allocate(alloc_size);
           alloc_cnt++;
+          if ((uint64_t)pmem_alloc->PMemUsageInBytes() >= pmem_size) break;
+          ASSERT_EQ(space_entry.size != 0, true);
         }
-        ASSERT_EQ(kvpairs, alloc_cnt);
-      };
-
-      LaunchNThreads(num_thread, TestPmemAlloc);
-      delete pmem_alloc;
+        ASSERT_EQ(pmem_size / block_size, alloc_cnt);
+        delete pmem_alloc;
+        TearDown();
```

Comment:
TearDown() is the destructor of the test case. You should never call it manually.
Write another function to remove pmem file.

Reply:
Done